samples/nrf9160/gnss/src/assistance.c

```diff
@@ -0,0 +1,340 @@
+	err = modem_info_string_get(MODEM_INFO_OPERATOR,
+				    resp_buf_cops,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err < 0) {
+		return err;
+	}
+
+	serving_cell->id = strtol(resp_buf_cellid, NULL, 16);
+	serving_cell->tac = strtol(resp_buf_tac, NULL, 16);
+	serving_cell->mnc = strtol(&resp_buf_cops[3], NULL, 10);
+	/* Null-terminate MCC, read and store it. */
+	resp_buf_cops[3] = '\0';
+	serving_cell->mcc = strtol(resp_buf_cops, NULL, 10);
+
+	return 0;
+}
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void get_pgps_data_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	int err;
+
+	printk("Sending request for P-GPS predictions to nRF Cloud...\n");
+
+	err = nrf_cloud_jwt_generate(0, jwt_buf, sizeof(jwt_buf));
+	if (err) {
+		printk("Failed to generate JWT, error: %d\n", err);
+
+		return;
+	}
+
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.timeout_ms = CONFIG_NRF_CLOUD_REST_RECV_TIMEOUT * MSEC_PER_SEC,
+		.auth = jwt_buf,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0, /* use CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE */
+		.status = 0,
+		.response = NULL,
+		.response_len = 0,
+		.total_response_len = 0
+	};
+
+	struct nrf_cloud_rest_pgps_request request = {
+		.pgps_req = &pgps_request
+	};
+
+	err = nrf_cloud_rest_pgps_data_get(&rest_ctx, &request);
+	if (err) {
+		printk("Failed to send P-GPS request, error: %d\n", err);
+
+		nrf_cloud_pgps_request_reset();
+
+		return;
+	}
+
+	printk("Processing P-GPS response\n");
+
+	err = nrf_cloud_pgps_process(rest_ctx.response, rest_ctx.response_len);
+	if (err) {
+		printk("Failed to process P-GPS response, error: %d\n", err);
+
+		nrf_cloud_pgps_request_reset();
+
+		return;
+	}
+
+	printk("P-GPS response processed\n");
+}
+
+static void inject_pgps_data_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	int err;
+
+	printk("Injecting P-GPS ephemerides\n");
+
+	err = nrf_cloud_pgps_inject(prediction, &agps_need);
+	if (err) {
+		printk("Failed to inject P-GPS ephemerides\n");
+	}
+
+	err = nrf_cloud_pgps_preemptive_updates();
+	if (err) {
+		printk("Failed to request P-GPS updates\n");
+	}
+}
+
+static void pgps_event_handler(struct nrf_cloud_pgps_event *event)
+{
+	switch (event->type) {
+	case PGPS_EVT_AVAILABLE:
+		prediction = event->prediction;
+
+		k_work_submit_to_queue(work_q, &inject_pgps_data_work);
+		break;
+
+	case PGPS_EVT_REQUEST:
+		assistance_active = true;
+
+		memcpy(&pgps_request, event->request, sizeof(pgps_request));
+
+		k_work_submit_to_queue(work_q, &get_pgps_data_work);
+		break;
+
+	case PGPS_EVT_LOADING:
+		printk("Loading P-GPS predictions\n");
+		assistance_active = true;
+		break;
+
+	case PGPS_EVT_READY:
+		printk("P-GPS predictions ready\n");
+		assistance_active = false;
+		break;
+
+	default:
+		/* No action needed */
+		break;
+	}
+}
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
```

Comment:
Same define ends and begins again, remove these lines

Reply:
True. I'll also move the semaphore definition to the beginning of the file and make it static.